torch_glow/src/GlowIValue.cpp

```diff
@@ -0,0 +1,281 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "GlowIValue.h"
+
+#include <ATen/core/ivalue.h>
+
+#include "glow/Base/Tensor.h"
+#include "glow/Support/Error.h"
+#include "glow/Support/Support.h"
+
+namespace glow {
+
+namespace {
+/// \returns a Glow tensor with the same type and underlying data as the given
+/// PyTorch tensor \p ptT.
+// TODO: don't redefine this
+llvm::Expected<glow::Tensor> ptTensorToGlowTensor2(const at::Tensor &ptT) {
+  // TODO: get correct ElemKind.
+  DCHECK_EQ(ptT.scalar_type(), at::kFloat)
+      << "Only float tensors supported currently.";
+  std::vector<size_t> dims;
+  for (const auto d : ptT.sizes()) {
+    dims.push_back(d);
+  }
+
+  glow::Type ty(glow::ElemKind::FloatTy, dims);
+  glow::Tensor glowT(ptT.data_ptr(), &ty);
+
+  return glowT;
+}
+
+const char *tagToStr(GlowIValue::Tag tag) {
+  switch (tag) {
+  case GlowIValue::Tag::None:
+    return "None";
+  case GlowIValue::Tag::Tensor:
+    return "Tensor";
+  case GlowIValue::Tag::Double:
+    return "Double";
+  case GlowIValue::Tag::Int:
+    return "Int";
+  case GlowIValue::Tag::Bool:
+    return "Bool";
+  case GlowIValue::Tag::IntList:
+    return "IntList";
+  case GlowIValue::Tag::DoubleList:
+    return "DoubleList";
+  case GlowIValue::Tag::BoolList:
+    return "BoolList";
+  case GlowIValue::Tag::TensorList:
+    return "TensorList";
+  case GlowIValue::Tag::Tuple:
+    return "Tuple";
+  }
+}
+} // namespace
+
+void GlowIValue::reset() {
+  switch (tag_) {
+  case Tag::Tensor:
+    delete payload_.asTensor;
+    break;
+  case Tag::IntList:
+    delete payload_.asIntList;
+    break;
+  case Tag::DoubleList:
+    delete payload_.asDoubleList;
+    break;
+  case Tag::BoolList:
+    delete payload_.asBoolList;
+    break;
+  case Tag::TensorList:
+    delete payload_.asTensorList;
+    break;
+  case Tag::Tuple:
+    delete payload_.asTuple;
+    break;
+  case Tag::None:
+  case Tag::Double:
+  case Tag::Int:
+  case Tag::Bool:
+    // Nothing to free.
+    break;
+  }
+  tag_ = Tag::None;
+}
+
+GlowIValue::~GlowIValue() { reset(); }
+
+GlowIValue::GlowIValue(GlowIValue &&other) {
+  std::swap(tag_, other.tag_);
+  std::swap(payload_, other.payload_);
+}
```

Comment:
Instead of swapping you could just set other.tag_=None.  It's kind of a pico-optimization though, I've just read [too many](https://www.quora.com/What-are-the-advantages-of-using-C-over-C-1/answer/Keith-Adams) [Scott Meyers articles](http://scottmeyers.blogspot.com/2014/06/the-drawbacks-of-implementing-move.html).

Reply:
Yeah that actually makes a lot of sense